### PR TITLE
gateway: handle max retries exceeded to return the last response message

### PR DESCRIPTION
**Pull request by shaotong-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/shaotong-db/mlflow/pull/10155?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10155/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10155
```

</p>
</details>

### What changes are proposed in this pull request?

This PR updates the http response handling logic in gateway. In case of max retries exceeded, no exceptions will be raised from python requests library and instead a response will be returned. In this way, the last server response can be fetched and displayed to the user.

We use `raise_on_status` in [urllib3.util.Retry](https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html) to achieve this. An unit test is added.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

When querying gateway with invalid data, such as `{"prompt": ["What is the best day of the week?"], "temperature": 0.3}` (notice `[]` should not exist here), the output looks like this:
```
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: xxx. Response text: {"error_code":"INTERNAL_ERROR","message":"Internal Error"}
``` 
instead of this (behavior before the change)
```
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='xxx', port=443): Max retries exceeded with url: xxx (Caused by ResponseError('too many 500 error responses'))
```

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/rest_utils.py [hunk 1]
```diff
@@ -108,6 +108,21 @@ def http_request(
...
         raise InvalidUrlException(f"Invalid url: {url}") from iu
+    except requests.exceptions.RetryError:
+        # When RetryError happens, make one more request to the server (without retries) and
+        # return the server's response to the caller. Displaying a more detailed error message
+        # is helpful for users to debug issues.
+        return _get_http_response_with_retries(
+            method,
+            url,
+            0,
+            backoff_factor,
```
Comment (BenWilson2) [inline mlflow/utils/rest_utils.py:119]: Do we need a backoff factor or can we just instantly resend and capture the failure by disabling the geometric backoff logic (setting the value to 0)?

File: mlflow/utils/rest_utils.py [hunk 2]
```diff
@@ -108,6 +108,21 @@ def http_request(
         ) from to
     except requests.exceptions.InvalidURL as iu:
         raise InvalidUrlException(f"Invalid url: {url}") from iu
+    except requests.exceptions.RetryError:
```
Comment (BenWilson2) [inline mlflow/utils/rest_utils.py:111]: This utility is used extensively in MLflow. Can we validate that this change has no adverse impacts to the client->tracking server in both OSS and Databricks and add client API tests that validate that there are no issues with current error responses associated with a retry-exhaustion failure mechanism for the tracking client?

Comment (shaotong-db) [reply mlflow/utils/rest_utils.py]: Thanks Ben! I had a offline discussion with @harupy and he suggested we could actually use `raise_on_status` field in the `Retry` config to handle this: https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html. The default value is `True` and I set it to `False` in gateway HTTP calls.

In this way, only the gateway caller behavior is changed while the other callers of `http_request` remain the same, thus the impact is minimized here.

File: mlflow/utils/rest_utils.py (hunk 1 above)
Comment (shaotong-db) [reply mlflow/utils/rest_utils.py]: This change is not relevant anymore. Please see the comment below :)

File: mlflow/utils/rest_utils.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/utils/rest_utils.py]: That is a FANTASTIC idea :D huge +1 to limiting the scope of the change only the gateway interface.

**Review by BenWilson2 — APPROVED** [review]
LGTM!

File: mlflow/utils/request_utils.py [hunk 3]
```diff
@@ -81,21 +82,22 @@ def _cached_get_request_session(
...
+        "raise_on_status": raise_on_status,
     }
     if Version(urllib3.__version__) >= Version("1.26.0"):
         retry_kwargs["allowed_methods"] = None
     else:
         retry_kwargs["method_whitelist"] = None
-
     retry = Retry(**retry_kwargs)
     adapter = HTTPAdapter(max_retries=retry)
+
```
Comment (harupy) [inline mlflow/utils/request_utils.py:93]: let's revert unrelated changes

**Review by harupy — APPROVED** [review]
LGTM!

File: tests/gateway/test_integration.py [hunk 4]
```diff
@@ -651,6 +651,36 @@ def _mock_request_session(max_retries, backoff_factor, retry_codes):
...
+    async def mock_chat(self, payload):
+        return expected_output
+
+    def _mock_request_session(max_retries, backoff_factor, retry_codes, raise_on_status):
+        return _cached_get_request_session(0, 1, retry_codes, False, os.getpid())
+
+    with patch(
+        "mlflow.utils.request_utils._get_request_session", _mock_request_session
+    ), patch.object(OpenAIProvider, "chat", mock_chat), pytest.raises(
+        requests.exceptions.HTTPError, match=".*Internal Server Error.*"
```
Comment (harupy) [inline tests/gateway/test_integration.py:679]: where does `Internal Server Error` come from?

Comment (shaotong-db) [reply tests/gateway/test_integration.py:679]: It comes from databricks gateway backend. If the behavior is changed there later (e.g. doesn't return `500` anymore), we can update the unit tests here.
